lib/policy_machine_storage_adapters/active_record/postgresql.rb

```diff
@@ -13,34 +12,86 @@ def self.transitive_closure?(ancestor, descendant)
-        else
-          PolicyElement.where('"policy_elements"."id" IN (SELECT assignments__recursive.parent_id FROM (WITH RECURSIVE "assignments__recursive" AS ( SELECT "assignments"."id", "assignments"."parent_id", "assignments"."child_id" FROM "assignments" WHERE "assignments"."child_id" in (?) UNION ALL SELECT "assignments"."id", "assignments"."parent_id", "assignments"."child_id" FROM "assignments" INNER JOIN "assignments__recursive" ON "assignments__recursive"."parent_id" = "assignments"."child_id" ) SELECT "assignments__recursive".* FROM "assignments__recursive") AS "assignments__recursive")', element_or_scope.map(&:id))
-        end
+        PolicyElement.where(
+          'id IN (
+            WITH RECURSIVE assignments_recursive AS (
+              (
+                SELECT parent_id, child_id
+                FROM assignments
+                WHERE child_id IN (?)
+              )
+              UNION ALL
+              (
+                SELECT assignments.parent_id, assignments.child_id
+                FROM assignments
+                INNER JOIN assignments_recursive
+                ON assignments_recursive.parent_id = assignments.child_id
+              )
+            )
+
+            SELECT assignments_recursive.parent_id
+            FROM assignments_recursive
+          )',
+          [*element_or_scope].map(&:id)
+        )
       end
 
+      # Returns the operation set IDs from the given list where the operation is
+      # a descendant of the operation set.
+      # TODO: Generalize this so that we can arbitrarily filter recursive assignments calls.
+      def self.filter_operation_set_list_by_assigned_operation(operation_set_ids, operation_id)
+        query = <<-SQL
+          WITH RECURSIVE assignments_recursive AS (
+            (
+              SELECT parent_id, child_id, ARRAY[parent_id] AS parents
+              FROM assignments
+              WHERE parent_id IN (#{operation_set_ids.join(',')})
+            )
+            UNION ALL
+            (
+              SELECT assignments.parent_id, assignments.child_id, assignments.parent_id || parents
+              FROM assignments
+              INNER JOIN assignments_recursive
+              ON assignments_recursive.child_id = assignments.parent_id
+            )
+          )
+
+          SELECT parents[1]
+          FROM assignments_recursive
+          JOIN policy_elements
+          ON policy_elements.id = assignments_recursive.child_id
+          WHERE policy_elements.unique_identifier = '#{operation_id}'
+          AND type = 'PolicyMachineStorageAdapter::ActiveRecord::Operation'
+        SQL
+        PolicyElement.connection.exec_query(query).rows.flatten.map(&:to_i)
```

Comment:
why `exec_query` instead of `execute` btw? also, is the result set here going to be big enough that we care about optimizing the `flatten.map` part?